File: backend/app/api/v1/assets.py
import logging

from fastapi import APIRouter, Query, status

from app.api.deps import CurrentUser, DBDep, require_permission
from app.core.permissions import Permissions, get_admin_scope
from app.schemas.asset import (
    AssetAssignRequest,
    AssetCategoryCreate,
    AssetCategoryOut,
    AssetCategoryUpdate,
    AssetCreate,
    AssetDetailOut,
    AssetOut,
    AssetUpdate,
    PaginatedAssets,
)
from app.services.asset import AssetService

router = APIRouter(tags=["assets"])
cat_router = APIRouter(prefix="/asset-categories", tags=["asset-categories"])
asset_router = APIRouter(prefix="/assets", tags=["assets"])

logger = logging.getLogger(__name__)


# ── Categories ──

@cat_router.get("", response_model=list[AssetCategoryOut])
async def list_categories(db: DBDep) -> list[AssetCategoryOut]:
    service = AssetService(db)
    categories = await service.list_categories()
    return [_build_category_tree(c, categories) for c in categories if c.parent_id is None]


def _build_category_tree(cat, all_cats: list) -> AssetCategoryOut:
    children = [
        _build_category_tree(c, all_cats)
        for c in all_cats
        if c.parent_id == cat.id
    ]
    return AssetCategoryOut(
        id=cat.id, name=cat.name, parent_id=cat.parent_id,
        description=cat.description, sort_order=cat.sort_order,
        created_at=cat.created_at, children=children,
    )


@cat_router.post("", response_model=AssetCategoryOut, status_code=status.HTTP_201_CREATED)
async def create_category(
    data: AssetCategoryCreate,
    db: DBDep,
    _perm: None = require_permission(Permissions.ASSET_CREATE),
) -> AssetCategoryOut:
    service = AssetService(db)
    cat = await service.create_category(data)
    return AssetCategoryOut.model_validate(cat)


@cat_router.put("/{category_id}", response_model=AssetCategoryOut)
async def update_category(
    category_id: int,
    data: AssetCategoryUpdate,
    db: DBDep,
    _perm: None = require_permission(Permissions.ASSET_UPDATE),
) -> AssetCategoryOut:
    service = AssetService(db)
    cat = await service.update_category(category_id, data)
    return AssetCategoryOut.model_validate(cat)


@cat_router.delete("/{category_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_category(
    category_id: int,
    db: DBDep,
    _perm: None = require_permission(Permissions.ASSET_DELETE),
) -> None:
    service = AssetService(db)
    await service.delete_category(category_id)


# ── Assets ──

@asset_router.get("/my", response_model=list[AssetOut])
async def list_my_assets(db: DBDep, current_user: CurrentUser) -> list[AssetOut]:
    service = AssetService(db)
    assets = await service.list_my_assets(current_user.id)
    return [AssetOut.model_validate(a) for a in assets]


@asset_router.get("", response_model=PaginatedAssets)
async def list_assets(
    db: DBDep,
    current_user: CurrentUser,
    _perm: None = require_permission(Permissions.ASSET_READ),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    search: str | None = None,
    category_id: int | None = None,
    status: str | None = None,
    department_id: int | None = None,
) -> dict:
    scope = get_admin_scope(current_user)
    if scope.scope == "department" and scope.dept_id:
        department_id = scope.dept_id
    service = AssetService(db)
    items, total = await service.list_assets(
        page=page, page_size=page_size, search=search,
        category_id=category_id, status=status, department_id=department_id,
    )
    return {
        "items": [AssetOut.model_validate(a) for a in items],
        "total": total,
        "page": page,
        "page_size": page_size,
    }


@asset_router.post("", response_model=AssetOut, status_code=status.HTTP_201_CREATED)
async def create_asset(
    data: AssetCreate,
    db: DBDep,
    _perm: None = require_permission(Permissions.ASSET_CREATE),
) -> AssetOut:
    service = AssetService(db)
    asset = await service.create_asset(data)
    return AssetOut.model_validate(asset)


@asset_router.get("/{asset_id}", response_model=AssetDetailOut)
async def get_asset(
    asset_id: int,
    db: DBDep,
    _perm: None = require_permission(Permissions.ASSET_READ),
) -> AssetDetailOut:
    service = AssetService(db)
    asset = await service.get_asset(asset_id)
    return AssetDetailOut.model_validate(asset)


@asset_router.put("/{asset_id}", response_model=AssetOut)
async def update_asset(
    asset_id: int,
    data: AssetUpdate,
    db: DBDep,
    _perm: None = require_permission(Permissions.ASSET_UPDATE),
) -> AssetOut:
    service = AssetService(db)
    asset = await service.update_asset(asset_id, data)
    return AssetOut.model_validate(asset)


@asset_router.delete("/{asset_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_asset(
    asset_id: int,
    db: DBDep,
    _perm: None = require_permission(Permissions.ASSET_DELETE),
) -> None:
    service = AssetService(db)
    await service.delete_asset(asset_id)


@asset_router.post("/{asset_id}/assign", response_model=AssetDetailOut)
async def assign_asset(
    asset_id: int,
    data: AssetAssignRequest,
    db: DBDep,
    current_user: CurrentUser,
    _perm: None = require_permission(Permissions.ASSET_UPDATE),
) -> AssetDetailOut:
    service = AssetService(db)
    asset = await service.assign(asset_id, data.user_id, current_user.id)
    return AssetDetailOut.model_validate(asset)


@asset_router.post("/{asset_id}/return", response_model=AssetDetailOut)
async def return_asset(
    asset_id: int,
    db: DBDep,
    current_user: CurrentUser,
    _perm: None = require_permission(Permissions.ASSET_UPDATE),
) -> AssetDetailOut:
    service = AssetService(db)
    asset = await service.return_asset(asset_id, current_user.id)
    return AssetDetailOut.model_validate(asset)

File: frontend/src/api/assets.ts
import api from './client'

export interface AssetCategory {
  id: number
  name: string
  parent_id: number | null
  description: string | null
  sort_order: number
  created_at: string
  children: AssetCategory[]
}

export interface AssetItem {
  id: number
  name: string
  category_id: number
  asset_code: string
  status: string
  department_id: number | null
  current_user_id: number | null
  purchase_date: string | null
  purchase_price: number | null
  supplier: string | null
  specification: Record<string, unknown> | null
  description: string | null
  created_at: string
  updated_at: string
  category: AssetCategory | null
  department: Record<string, unknown> | null
}

export interface AssetDetail extends AssetItem {
  current_user: Record<string, unknown> | null
  assignments: AssetAssignment[]
}

export interface AssetAssignment {
  id: number
  asset_id: number
  user_id: number
  action: string
  action_date: string
  notes: string | null
  operator_id: number
  created_at: string
  user: Record<string, unknown> | null
  operator: Record<string, unknown> | null
}

export interface PaginatedAssets {
  items: AssetItem[]
  total: number
  page: number
  page_size: number
}

export const assetStatusLabel = (s: string) => {
  const map: Record<string, string> = {
    in_use: '使用中', idle: '闲置', scrapped: '已报废', repairing: '维修中',
  }
  return map[s] || s
}

export const assetStatusColor = (s: string) => {
  const map: Record<string, string> = {
    in_use: 'text-green-600', idle: 'text-gray-400', scrapped: 'text-red-500', repairing: 'text-yellow-600',
  }
  return map[s] || ''
}

export const assetApi = {
  listCategories: () => api.get<AssetCategory[]>('/asset-categories'),

  createCategory: (data: { name: string; parent_id?: number; description?: string; sort_order?: number }) =>
    api.post<AssetCategory>('/asset-categories', data),

  updateCategory: (id: number, data: Record<string, unknown>) =>
    api.put<AssetCategory>(`/asset-categories/${id}`, data),

  deleteCategory: (id: number) =>
    api.delete(`/asset-categories/${id}`),

  list: (params?: Record<string, unknown>) =>
    api.get<PaginatedAssets>('/assets', { params }),

  listMy: () =>
    api.get<AssetItem[]>('/assets/my'),

  getById: (id: number) =>
    api.get<AssetDetail>(`/assets/${id}`),

  create: (data: Record<string, unknown>) =>
    api.post<AssetItem>('/assets', data),

  update: (id: number, data: Record<string, unknown>) =>
    api.put<AssetItem>(`/assets/${id}`, data),

  delete: (id: number) =>
    api.delete(`/assets/${id}`),

  assign: (id: number, userId: number) =>
    api.post<AssetDetail>(`/assets/${id}/assign`, { user_id: userId }),

  return: (id: number) =>
    api.post<AssetDetail>(`/assets/${id}/return`),
}
